bus-mapping/src/circuit_input_builder.rs

```diff
@@ -577,62 +645,205 @@ impl<'a> CircuitInputStateRef<'a> {
+                (call_data.0, call_data.1, return_data.0, return_data.1)
+            }
+            CallKind::DelegateCall | CallKind::StaticCall => {
+                let call_data = get_memory_offset_length(2)?;
+                let return_data = get_memory_offset_length(4)?;
+                (call_data.0, call_data.1, return_data.0, return_data.1)
             }
+            CallKind::Create | CallKind::Create2 => (0, 0, 0, 0),
         };
-        self.push_call(kind, address, code_source, code_hash);
+
+        let caller = &self.tx.calls[self.tx_ctx.call_index()];
+        let call = Call {
+            call_id: self.block_ctx.rwc.0,
+            kind,
+            is_static: kind == CallKind::StaticCall || caller.is_static,
+            is_root: false,
+            is_persistent: caller.is_persistent && is_success,
+            is_success,
+            rw_counter_end_of_reversion: 0,
+            caller_address,
+            address,
+            code_source,
+            code_hash,
+            depth: caller.depth + 1,
+            value,
+            call_data_offset,
+            call_data_length,
+            return_data_offset,
+            return_data_length,
+        };
+        self.push_call(call);
+
         Ok(())
     }
 
     /// Handle a return step caused by any opcode that causes a return to the
     /// previous call context.
-    fn handle_return(
-        &mut self,
-        step: &GethExecStep,
-        next_step: &GethExecStep,
-    ) -> Result<(), Error> {
-        if self.tx_ctx.call_stack.len() == 1 {
-            return Err(Error::InvalidGethExecStep(
-                "handle_tx: call stack will be empty",
-                Box::new(step.clone()),
-            ));
-        }
-        let (_, call_ctx) = self
-            .tx_ctx
-            .pop_call_index_ctx()
-            .expect("call stack is empty");
-        // If the return was successful, accumulate the swc from the
-        // subcall.
-        if !next_step.stack.last()?.is_zero() {
-            self.tx_ctx.call_ctx_mut().swc += call_ctx.swc;
+    fn handle_return(&mut self, is_success: bool) -> Result<(), Error> {
+        let (call_idx, call_ctx) = self.tx_ctx.pop_call_index_ctx().ok_or(
+            Error::InvalidGethExecTrace("handle_return: call stack is empty"),
+        )?;
+        if is_success {
+            // If the return was successful, accumulate the success callees and
+            // swc.
+            if let Some((_, caller_ctx)) = self.tx_ctx.call_stack.last_mut() {
+                caller_ctx.success_callees.extend(call_ctx.success_callees);
+                caller_ctx.success_callees.push((call_idx, caller_ctx.swc));
+                caller_ctx.swc += call_ctx.swc;
+                caller_ctx.operations.extend(call_ctx.operations);
+            }
+        } else {
+            // Apply state reversion and push the operations.
+            for op in call_ctx.operations.into_iter().rev() {
```

Comment:
1. Yes, more precisely, currently we only call `push_op_reverted` when it's a write.
2. Yes, we should swap the value, currently it's swapped manually... I think it should be automatically done in some way just like the spec. I tried but it requires much more changes on `CircuitInputBuilder`.

Reply:
Ok, so if I got it correctly it goes like this:
In `CallContext`, `operations` accumulates the inverse of every state operation done in the call.
If the return is successful, all these operations are accumulated in the caller context.
If the return is failure, all these operations are iterated in reverse where:
- the change in the state_db, access_list is undone (by applying the reverse operation)
- the reverse ops are pushed to the return step (this assigns the rwc)

Regarding point 2.2: You mean that the `OperationRef` of the reverse operation belongs to the step that generated the original operation, but there's no need to set them in the `CircuitInputBuilder` because the circuit code can easily infer that?  And why is the `OperationRef` of the reverse operation assigned to the return step?

Reply:
> Also, If I understand this correctly, with this approach, we should generate the reverse operation and do push_op_reverted in every Opcode's gen_associated_ops for every state write right?

Yes, and I think after we can handle call/create in `gen_associated_ops`, we can make it done automatically inside some function, because we then don't need to explicitly assign which context we are working on.

> Regarding point 2.2: You mean that the OperationRef of the reverse operation belongs to the step that generated the original operation, but there's no need to set them in the CircuitInputBuilder because the circuit code can easily infer that?

Yes, but for completeness, I think we still need to add fulfill it.

> And why is the OperationRef of the reverse operation assigned to the return step?

Ah, I should not call `push_op` since it pushes the `OperationRef` into current step. I only intent to insert the operation into block. Let me fix it.